<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        {
            function p1() {
                setTimeout(() => {
                    let a = 1
                    setTimeout(() => {
                        let b = a + 2;
                        setTimeout(() => {
                            console.log(b + 3);
                        }, 1000)
                    }, 1000)
                }, 1000)
            }
            p1()
        }
        {
            const p = new Promise((resolve, reject) => { resolve('resolve'); reject('reject') });
            console.log(p);
        }

        {

        }

        // const p1 = new Promise(() => setTimeout(console.log, 0, 'executor'));
        // setTimeout(console.log, 0, 'promise');
        // setTimeout(console.log, 0, p1);  // Promise {<pending>}

        // const p2 = new Promise((res, rej) => res());
        // 已解决的期约
        // const p3 = Promise.resolve();

        // promise.resolve()传入多个值，除第一个外的其他值都会被忽略，
        // setTimeout(console.log, 0, p3); // Promise {<fulfilled>: undefined}
        // setTimeout(console.log, 0, Promise.resolve(222)); // Promise {<fulfilled>: 222}
        // setTimeout(console.log, 0, Promise.resolve(1, 3, 4, 5)); // Promise {<fulfilled>: 1}

        // 已解决期约传入一个期约，类似于一个空包装，状态表现为传入期约的状态
        // const p4 = new Promise(() => { });
        // setTimeout(console.log, 0, p4 === Promise.resolve(p4)); // true
        // setTimeout(console.log, 0, Promise.resolve(p4)); // Promise {<pending>}
        // setTimeout(console.log, 0, p4); // Promise {<pending>}
        // const p5 = Promise.resolve(4);
        // setTimeout(console.log, 0, Promise.resolve(p5) === p5);// true
        // setTimeout(console.log, 0, Promise.resolve(p5)); // Promise {<fulfilled>: 4}
        // setTimeout(console.log, 0, p5); // Promise {<fulfilled>: 4}
        // setTimeout(console.log, 0, Promise.resolve(Promise.reject('w'))); // Promise {<rejected>: "w"}

        // setTimeout(console.log, 0, Promise.resolve(4) === Promise.resolve(4));
        // 传入任何非期约值都会转换为已解决期约包括错误对象
        // console.log(Promise.resolve(new Error('error'))); // Promise {<fulfilled>: Error: error


        // Promise.reject
        // setTimeout(console.log, 0, Promise.reject()); // Promise {<rejected>: undefined}
        // setTimeout(console.log, 0, Promise.reject(222)); // Promise {<rejected>: 222}
        // setTimeout(console.log, 0, Promise.reject(1, 3, 4, 5)); // Promise {<rejected>: 1}
        // Promise.reject与Promise.resolve()不同，传入期约对象则这个期约对象会变成reject的值
        // setTimeout(console.log, 0, Promise.reject(Promise.resolve('wmy'))) //Promise {<rejected>: Promise{<resolved>: "wmy"}}
        // setTimeout(console.log, 0, Promise.reject(Promise.resolve())) // Promise <rejected>: Promise <resolved>
        // setTimeout(console.log, 0, Promise.reject(3)) // Promise {<rejected>: 3}

        // promise.prototype.catch
        // 是Promise.prototype.then(null, (res, rej) => {})的语法糖

        // 执行顺序
        // let p = Promise.resolve();
        // console.log(1);
        // p.then((res) => console.log(2));
        // console.log(3);

        // let synchronousResolve;
        // // 创建一个期约并将解决函数保存在一个局部变量中
        // let p = new Promise((resolve) => {
        //     synchronousResolve = function () {
        //         console.log('1: invoking resolve()');
        //         resolve();
        //         console.log('2: resolve() returns');
        //     };
        // });
        // p.then(() => console.log('4: then() handler executes'));
        // synchronousResolve();
        // console.log('3: synchronousResolve() returns');

        // 期约错误
        // console.log('error start');
        // throw new Error('foo');
        // console.log('error end');

        // console.log('eror start');
        // Promise.reject('foo');
        // // new Promise((resolve, reject) => reject('foo'))
        // console.log('error end');

        // try {
        //     throw Error('start');
        //     new Promise((res, rej) => {
        //         rej('reject error');
        //     })
        //         .catch(e => console.log('catch ' + e))
        //         .finally(() => {throw Error('finally error')})
        // } catch (e) {
        //     console.log('err' + e)
        // }

        let p1 = Promise.all([
            Promise.resolve(),
            Promise.resolve()
        ]);
        console.log(p1);
        // 可迭代对象中的元素会通过 Promise.resolve()转换为期约
        let p2 = Promise.all([3, 4]);
        console.log(p2);
        // 空的可迭代对象等价于 Promise.resolve()
        let p3 = Promise.all([]);
        console.log(p3);
        // 无效的语法
        // let p4 = Promise.all();
// TypeError: cannot read Symbol.iterator of undefined
    </script>
</body>

</html>